File: packages/evaluatorq/src/lib/integrations/simulation/wrap-agent.ts
/**
 * Wraps the simulation framework as an evaluatorq Job.
 *
 * Follows the same pattern as wrapAISdkAgent() and wrapLangChainAgent().
 */

import type { DataPoint, Job, Output } from "../../types.js";
import { fromOrqDeployment } from "./adapters.js";
import { toOpenResponses } from "./convert.js";
import { simulate } from "./simulation/index.js";
import type {
  ChatMessage,
  Datapoint,
  Persona,
  Scenario,
  SimulationResult,
} from "./types.js";

/**
 * Options for creating a simulation job.
 */
export interface SimulationJobOptions {
  /** Display name for this job in results. Defaults to "simulation". */
  name?: string;
  /** Target agent callback — receives messages and returns a response string. */
  targetCallback?: (messages: ChatMessage[]) => string | Promise<string>;
  /** Orq deployment key — used if targetCallback is not provided. */
  agentKey?: string;
  /** Maximum conversation turns per simulation. Defaults to 10. */
  maxTurns?: number;
  /** Model used for user simulator and judge agents. Defaults to "azure/gpt-4o-mini". */
  model?: string;
  /** Built-in evaluator names to apply to results. Defaults to ["goal_achieved", "criteria_met"]. */
  evaluators?: string[];
}

/**
 * Creates an evaluatorq Job that runs agent simulations.
 *
 * Each DataPoint should have inputs containing simulation data:
 * - `persona` (Persona object) and `scenario` (Scenario object), or
 * - `datapoint` (full Datapoint object), or
 * - `personas` (Persona[]) and `scenarios` (Scenario[]) for batch generation
 *
 * The job:
 * 1. Extracts persona/scenario/datapoint from data.inputs
 * 2. Runs simulate() with the target agent
 * 3. Converts the first result to OpenResponses format
 * 4. Returns { name, output: ResponseResource }
 *
 * @example
 * ```typescript
 * import { wrapSimulationAgent } from "@orq-ai/evaluatorq/simulation";
 *
 * const job = wrapSimulationAgent({
 *   targetCallback: async (messages) => {
 *     // Your agent logic here
 *     return "Agent response";
 *   },
 *   maxTurns: 5,
 * });
 *
 * await evaluatorq("simulation-eval", {
 *   data: [
 *     {
 *       inputs: {
 *         persona: { name: "Impatient User", patience: 0.2, ... },
 *         scenario: { name: "Refund Request", goal: "Get a refund", ... },
 *       },
 *     },
 *   ],
 *   jobs: [job],
 * });
 * ```
 */
export function wrapSimulationAgent(options: SimulationJobOptions): Job {
  const {
    name = "simulation",
    targetCallback,
    agentKey,
    maxTurns = 10,
    model,
    evaluators,
  } = options;

  return async (data: DataPoint, _row: number) => {
    // Resolve the target callback
    let resolvedCallback = targetCallback;
    if (!resolvedCallback && agentKey) {
      resolvedCallback = fromOrqDeployment(agentKey);
    }

    if (!resolvedCallback) {
      throw new Error(
        "wrapSimulationAgent requires either targetCallback or agentKey",
      );
    }

    // Extract simulation inputs from DataPoint
    const inputs = data.inputs;

    let datapoints: Datapoint[] | undefined;
    let personas: Persona[] | undefined;
    let scenarios: Scenario[] | undefined;

    if (inputs.datapoint) {
      const dp = inputs.datapoint;
      validateShape(dp, "datapoint", ["persona", "scenario", "first_message"]);
      datapoints = [dp as Datapoint];
    } else if (inputs.datapoints) {
      const dps = inputs.datapoints;
      if (!Array.isArray(dps)) {
        throw new Error("Expected 'datapoints' to be an array");
      }
      for (const dp of dps) {
        validateShape(dp, "datapoints[]", [
          "persona",
          "scenario",
          "first_message",
        ]);
      }
      datapoints = dps as Datapoint[];
    } else if (inputs.persona && inputs.scenario) {
      validateShape(inputs.persona, "persona", ["name"]);
      validateShape(inputs.scenario, "scenario", ["name", "goal"]);
      personas = [inputs.persona as Persona];
      scenarios = [inputs.scenario as Scenario];
    } else if (inputs.personas && inputs.scenarios) {
      if (!Array.isArray(inputs.personas) || !Array.isArray(inputs.scenarios)) {
        throw new Error("Expected 'personas' and 'scenarios' to be arrays");
      }
      for (const p of inputs.personas) validateShape(p, "personas[]", ["name"]);
      for (const s of inputs.scenarios)
        validateShape(s, "scenarios[]", ["name", "goal"]);
      personas = inputs.personas as Persona[];
      scenarios = inputs.scenarios as Scenario[];
    } else {
      throw new Error(
        "Expected data.inputs to contain 'persona' + 'scenario', 'datapoint', 'datapoints', or 'personas' + 'scenarios'",
      );
    }

    // Run simulation
    const results: SimulationResult[] = await simulate({
      evaluationName: name,
      targetCallback: resolvedCallback,
      datapoints,
      personas,
      scenarios,
      maxTurns,
      model,
      evaluators,
    });

    // Convert first result to OpenResponses format
    const result = results[0];
    if (!result) {
      throw new Error("Simulation produced no results");
    }
    if (results.length > 1) {
      console.warn(
        `wrapSimulationAgent: ${results.length} simulations ran but only the first result is returned. ` +
          "Use simulate() directly to collect all results.",
      );
    }

    const openResponsesOutput = toOpenResponses(result, model);

    return {
      name,
      output: openResponsesOutput as unknown as Output,
    };
  };
}

/** Lightweight runtime check that an object has the expected keys. */
function validateShape(
  value: unknown,
  label: string,
  requiredKeys: string[],
): void {
  if (typeof value !== "object" || value === null) {
    throw new Error(`Expected '${label}' to be an object, got ${typeof value}`);
  }
  for (const key of requiredKeys) {
    if (!(key in value)) {
      throw new Error(`Invalid '${label}': missing required field '${key}'`);
    }
  }
}

File: packages/evaluatorq-py/src/evaluatorq/simulation/wrap_agent.py
"""Wraps the simulation framework as an evaluatorq Job.

Follows the same pattern as wrap_langchain_agent().
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any

from evaluatorq.simulation._datapoint_io import _extract_single_datapoint
from evaluatorq.simulation.adapters import from_orq_deployment
from evaluatorq.simulation.convert import to_open_responses
from evaluatorq.simulation.types import (
    DEFAULT_MODEL,
    ChatMessage,
)

if TYPE_CHECKING:
    from collections.abc import Awaitable, Callable

    from evaluatorq.simulation.agents.base import BaseAgent
    from evaluatorq.types import DataPoint

logger = logging.getLogger(__name__)


def wrap_simulation_agent(
    *,
    name: str = "simulation",
    target_callback: Callable[[list[ChatMessage]], str | Awaitable[str]] | None = None,
    agent_key: str | None = None,
    max_turns: int = 10,
    model: str | None = None,
    user_simulator: BaseAgent | None = None,
    judge: BaseAgent | None = None,
    **deprecated_kwargs: Any,
) -> Callable[[DataPoint, int], Awaitable[dict[str, Any]]]:
    """Create an evaluatorq Job that runs agent simulations.

    Each DataPoint should have inputs containing simulation data:
    - ``persona`` and ``scenario``, or
    - ``datapoint`` (full Datapoint object), or
    - ``datapoints`` / ``personas`` + ``scenarios`` each of length one

    The returned callable owns a long-lived ``SimulationRunner`` (and its
    underlying HTTP client). Call ``await job_fn.aclose()`` after your
    ``evaluatorq()`` run finishes to release the connection pool — otherwise
    it leaks until process exit. Example::

        job = wrap_simulation_agent(target_callback=cb)
        try:
            await evaluatorq("run", data=[...], jobs=[job], evaluators=[...])
        finally:
            await job.aclose()
    """
    from evaluatorq.simulation.runner.simulation import SimulationRunner

    if "evaluators" in deprecated_kwargs:
        # Removed in RES-594: scoring belongs on the evaluatorq() call, not
        # the job that produces the output. Raise loud — silently dropping
        # scoring would be worse than a TypeError.
        raise TypeError(
            "wrap_simulation_agent() no longer accepts 'evaluators='. Pass your "
            "evaluator list to evaluatorq(..., evaluators=...) instead. See CHANGELOG."
        )
    if deprecated_kwargs:
        raise TypeError(
            f"wrap_simulation_agent() got unexpected keyword arguments: "
            f"{sorted(deprecated_kwargs)}"
        )

    resolved_callback = target_callback
    if not resolved_callback and agent_key:
        resolved_callback = from_orq_deployment(agent_key)
    if not resolved_callback:
        raise ValueError(
            "wrap_simulation_agent requires either target_callback or agent_key"
        )

    effective_model = model or DEFAULT_MODEL

    runner = SimulationRunner(
        target_callback=resolved_callback,
        model=effective_model,
        max_turns=max_turns,
        user_simulator=user_simulator,
        judge=judge,
    )

    async def job_fn(data: DataPoint, _row: int) -> dict[str, Any]:
        sim_dp = _extract_single_datapoint(data)
        result = await runner.run(datapoint=sim_dp, max_turns=max_turns)
        return {
            "name": name,
            "output": to_open_responses(result, effective_model),
        }

    async def aclose() -> None:
        await runner.close()

    setattr(job_fn, "aclose", aclose)  # noqa: B010
    return job_fn
